lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2387,120 @@ def _containers(self) -> Dict[str, Container]:
         return {cont: self._charm.unit.get_container(cont) for cont in self._logs_scheme.keys()}
 
 
+class LogForwarder(Object):
+    """Forward the standard outputs of all workloads operated by a charm to one or multiple Loki endpoints."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        loki_endpoint_getter: Optional[Callable[[str], List[str]]] = None,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.topology = JujuTopology.from_charm(charm)
+        self.endpoints_getter: Callable = (
```

Comment:
this should be private IMHO. Also `topology`, and `loki_endpoints`

Take a close look at the public signature of this class and remove all that shouldn't be visible.